### PR TITLE
Fix Email field test

### LINKED ISSUE #574: Test FieldTypesTest::testEmail3 is failing
This is the only one test that requires DNS - DNS check function should be separated in extra method and this method should be overrided in tests.

### PR DESCRIPTION
**Pull request by DarkSide666**
fix #574

### CODE REVIEW

File: src/Field/Email.php [hunk 1]
```diff
@@ -84,14 +88,31 @@ public function normalize($value)
...
+     */
+    protected function isDNSValid(string $domain): bool
+    {
+        return $this->hasDNSRecord($domain, true) || $this->hasDNSRecord($domain, false);
+    }
+
+    private function hasDNSRecord(string $domain, bool $isMX): bool
+    {
+        $normalizedDomain = $domain . '.';
+        if (!checkdnsrr($normalizedDomain, ($isMX ? 'MX' : 'A'))) {
```
Comment (mvorisek) [inline src/Field/Email.php]: Do we need `checkdnsrr` at all?

File: src/Field/Email.php [hunk 2]
```diff
@@ -84,14 +88,31 @@ public function normalize($value)
...
+            return false;
+        }
+
+        // dns_get_record can also return false
+        $records = dns_get_record($normalizedDomain, ($isMX ? DNS_MX : DNS_A)) ?: [];
+
+        return !empty($records);
```
Comment (mvorisek) [inline src/Field/Email.php]: much better to test `!== false && count() > 0`, then you are do not need any comment ;-)

File: src/Field/Email.php [hunk 3]
```diff
@@ -84,14 +88,31 @@ public function normalize($value)
         return parent::normalize(implode(', ', $emails));
     }
 
+    /**
+     * Validate DNS.
+     */
+    protected function isDNSValid(string $domain): bool
```
Comment (mvorisek) [inline src/Field/Email.php]: What about one method like `hasAnyDnsRecord(string $domain, array $types = ['MX', 'A'])`?

Comment (DarkSide666) [reply src/Field/Email.php]: Can do like that but what it will change? :)

File: src/Field/Email.php (hunk 2 above)
Comment (DarkSide666) [reply src/Field/Email.php]: addressed

File: src/Field/Email.php (hunk 1 above)
Comment (DarkSide666) [reply src/Field/Email.php]: Not sure, but I guess it's good to check and step out of method early.
Probably get_dns_records() takes more time and resources to execute.

File: src/Field/Email.php (hunk 3 above)
Comment (mvorisek) [reply src/Field/Email.php]: A little cleanup, nothing more, no core functionality. I opened this mainly as this test was failing and occupying phpunit output as skipped :D

File: src/Field/Email.php [hunk 4]
```diff
@@ -103,7 +103,6 @@ private function hasDNSRecord(string $domain, bool $isMX): bool
             return false;
         }
 
-        // dns_get_record can also return false
         $records = dns_get_record($normalizedDomain, ($isMX ? DNS_MX : DNS_A)) ?: [];
 
         return $records !== false && count($records) > 0;
```
Comment (mvorisek) [inline src/Field/Email.php]: ` ?: []` - with this you do not need `!== false`...

Comment (DarkSide666) [reply src/Field/Email.php]: oh yes, right :)

File: tests/FieldTypesTest.php [hunk 5]
```diff
@@ -74,8 +79,9 @@ public function testEmail4()
 
         $m->set('email_name', 'Romans <me@gmail.com>');
         $m->set('email_names', 'Romans1 <me1@gmail.com>, Romans2 <me2@gmail.com>; Romans3 <me3@gmail.com>');
-        $m->set('email_idn', 'test@日本レジストリサービス.jp');
+        $m->set('email_idn', 'test@háčkyčárky.cz'); // official IDN test domain
```
Comment (mvorisek) [inline tests/FieldTypesTest.php:82]: not working with non-encoded IDN domain

File: src/Field/Email.php [hunk 6]
```diff
@@ -64,17 +64,20 @@ public function normalize($value)
...
-            $p = explode('@', $email, 2);
-            $user = $p[0] ?? null;
-            $domain = $p[1] ?? null;
-            if (!filter_var($user . '@' . $this->idn_to_ascii($domain), FILTER_VALIDATE_EMAIL)) {
-                throw new ValidationException([$this->name => 'Email format is invalid']);
+            if (strpos($email, '@') === false) {
+                throw new ValidationException([$this->name => 'Email address does not have domain']);
+            }
+
+            [$user, $domain] = explode('@', $email, 2);
```
Comment (mvorisek) [inline src/Field/Email.php:71]: `@` can still be within `<>` name, but ok for now...

File: src/Field/Email.php [hunk 7]
```diff
@@ -64,17 +64,20 @@ public function normalize($value)
...
+            $domain = idn_to_ascii($domain); // always convert domain to ASCII
```
Comment (mvorisek) [inline src/Field/Email.php:72]: makes sense to store in ASCII for DB normalization

File: src/Field/Email.php [hunk 8]
```diff
@@ -84,15 +87,22 @@ public function normalize($value)
...
-        return idn_to_ascii($domain, IDNA_NONTRANSITIONAL_TO_ASCII, INTL_IDNA_VARIANT_UTS46);
+        foreach (array_unique(array_map('strtoupper', $types)) as $t) {
+            $dnsConsts = [
+                'MX' => DNS_MX,
+                'A' => DNS_A,
+                'AAAA' => DNS_AAAA,
+                'CNAME' => DNS_CNAME,
+            ];
+
+            $records = dns_get_record($domain . '.', $dnsConsts[$t]);
```
Comment (mvorisek) [inline src/Field/Email.php:100]: this seems to always work

File: tests/FieldTypesTest.php (hunk 5 above)
Comment (mvorisek) [reply tests/FieldTypesTest.php:82]: was a bug in your code, you checked with ascii domain only in format

Comment (DarkSide666) [reply tests/FieldTypesTest.php:82]: Romans Yellow Horse :) :) :)

**Pull request merged**